lib/lte_link_control/lte_lc_helpers.c

```diff
@@ -435,6 +435,147 @@ int parse_edrx(const char *at_response, struct lte_lc_edrx_cfg *cfg)
+	uint32_t requested_value,
+	const uint32_t *lookup_table,
+	uint8_t lookup_table_size)
+{
+	/* Timer unit and timer value refer to the terminology used in 3GPP 24.008
+	 * Ch. 10.5.7.4a and Ch. 10.5.7.3. for bits 6 to 8 and bits 1 to 5, respectively
+	 */
+
+	/* Lookup table index to the currently selected timer unit */
+	uint32_t selected_index = -1;
+	/* Currently calculated value for the timer, which tries to match the requested value */
+	uint32_t selected_value = -1;
+	/* Currently selected timer value */
+	uint32_t selected_timer_value = -1;
+	/* Timer unit and timer value encoded as an integer which will be converted to a string */
+	uint8_t encoded_value_int = 0;
+
+	/* Search a value that is as close as possible to the requested value
+	 * rounding it up to the closest possible value
+	 */
+	for (int i = 0; i < lookup_table_size; i++) {
+		uint32_t current_timer_value = requested_value / lookup_table[i];
+
+		if (requested_value % lookup_table[i] > 0) {
+			/* Round up the time when it's not divisible by the timer unit */
+			current_timer_value++;
+		}
+
+		/* Current timer unit is so small that timer value range is not enough */
+		if (current_timer_value > TIMER_VALUE_MAX) {
+			continue;
+		}
+
+		uint32_t current_value = lookup_table[i] * current_timer_value;
+
+		/* Use current timer unit if current value is closer to requested value
+		 * than currently selected values
+		 */
+		if (selected_value == -1 ||
+		    current_value - requested_value  < selected_value - requested_value) {
+			selected_value = current_value;
+			selected_index = i;
+			selected_timer_value = current_timer_value;
+		}
+	}
+
+	if (selected_index != -1) {
+		LOG_DBG("requested_value=%d, selected_value=%d, selected_timer_unit=%d, "
+			"selected_timer_value=%d, selected_index=%d",
+			requested_value,
+			selected_value,
+			lookup_table[selected_index],
+			selected_timer_value,
+			selected_index);
+
+		/* Selected index (timer unit) is in bits 6 to 8 */
+		encoded_value_int = (selected_index << 5) | selected_timer_value;
+		sprintf(encoded_timer_str, BYTE_TO_BINARY_STRING_FORMAT,
+			BYTE_TO_BINARY_STRING_ARGS(encoded_value_int));
+	} else {
+		LOG_ERR("requested_value=%d is too big to be represented by the timer encoding",
+			requested_value);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+int encode_psm(char *tau_ext_str, char *active_time_str, int rptau, int rat)
+{
+	int ret = 0;
+
+	LOG_DBG("TAU: %d sec, active time: %d sec", rptau, rat);
+
+	__ASSERT_NO_MSG(tau_ext_str != NULL);
+	__ASSERT_NO_MSG(active_time_str != NULL);
+
+	if (rptau >= 0) {
+		ret = encode_psm_timer(
+			tau_ext_str,
+			rptau,
+			t3412_ext_lookup,
+			T3412_EXT_LOOKUP_DIFFERENT_VALUES);
+	} else {
+		*tau_ext_str = '\0';
+		LOG_DBG("RPTAU use default");
+	}
+
+	if (rat >= 0) {
+		ret |= encode_psm_timer(
+			active_time_str,
+			rat,
+			t3324_lookup,
+			T3324_LOOKUP_DIFFERENT_VALUES);
+	} else {
+		*active_time_str = '\0';
+		LOG_DBG("RPTAU use default");
```

Comment:
Should be RAT, not RPTAU.
```suggestion
		LOG_DBG("Using modem default value for RAT");
```